File: FLIR/09_Trigger.py
#!/usr/bin/env python3

"""
Shows how to use the Trigger feature:
The camera only take an image when the trigger event happens.
When using SoftwareTrigger, SoftwareTrigger needs to be called.
Blackfly camera also support other triggers e.g. GPIO Input.

Reference: Acquisition Control section in Technical Reference.

Date: 2019-12-31
"""

import PySpin


system = PySpin.System_GetInstance()

cam_list = system.GetCameras()
cam = cam_list.GetByIndex(0)
cam.Init()


# turn off trigger mode before configuration
cam.TriggerMode.SetValue(PySpin.TriggerMode_Off)
cam.TriggerSource.SetValue(PySpin.TriggerSource_Software)
cam.TriggerMode.SetValue(PySpin.TriggerMode_On)
cam.AcquisitionMode.SetValue(PySpin.AcquisitionMode_Continuous)

cam.BeginAcquisition()
try:
    img = cam.GetNextImage(1000)
    img.Release()
except PySpin.SpinnakerException as e:
    print('Expected timeout, no image taken yet: {}'.format(e))

cam.TriggerSoftware.Execute()
img = cam.GetNextImage(1000)
print('One image taken FrameID: {}'.format(img.GetFrameID()))
img.Release()

cam.EndAcquisition()

# reset trigger mode to off
cam.TriggerMode.SetValue(PySpin.TriggerMode_Off)

cam.DeInit()
del cam
cam_list.Clear()
system.ReleaseInstance()

File: FLIR/09_Trigger.cpp
/*
 * Shows how to use trigger feature.
 *
 * */

#include "Spinnaker.h"
#include <iostream>

using namespace Spinnaker;
using namespace Spinnaker::GenApi;
using namespace Spinnaker::GenICam;


int main(int /*argc*/, char** /*argv*/)
{
    std::cout << "FLIR: Acquisition\n";
    auto system = System::GetInstance();
    auto cam_list = system->GetCameras();

    if (cam_list.GetSize() == 0) {
        std::cerr << "No cameras found\n";
        cam_list.Clear();
        system->ReleaseInstance();
        return -1;
    }

    auto cam = cam_list.GetByIndex(0);
    cam->Init();
    std::cout << "Fancy processing\n";
    cam->DeInit();

    // turn off trigger mode
    cam->TriggerMode.SetValue(TriggerMode_Off);
    cam->TriggerSource.SetValue(TriggerSource_Software);
    cam->TriggerMode.SetValue(TriggerMode_On);
    cam->AcquisitionMode.SetValue(AcquisitionMode_SingleFrame);

    cam->BeginAcquisition();

    cam->TriggerSoftware.Execute();
    auto img = cam->GetNextImage(1000);
    uint64_t timestamp = img->GetTimeStamp();
    img->Release();
    cam->EndAcquisition();

    // reset trigger mode to off
    cam->TriggerMode.SetValue(TriggerMode_Off);

    // Release reference to the camera
    cam = nullptr;
    cam_list.Clear();
    system->ReleaseInstance();
    return 0;
}
